Allow two-way bindings to signal-based template variables

crisbeto opened the pull request:
Includes the following changes whose goal is to properly support two-way bindings to signal-based local variables:

### refactor(compiler-cli): introduce template semantics checker
Introduces a new `TemplateSemanticsChecker` that will be used to flag semantic errors in the user's template. Currently we do some of this in the type check block, but the problem is that it doesn't have access to the template type checker which prevents us from properly checking cases like https://github.com/angular/angular/issues/54670. This pass is also distinct from the extended template checks, because we don't want users to be able to turn the checks off and we want them to run even if `strictTemplates` are disabled.

### refactor(compiler-cli): move signal identification function
Moves the function that identifies signals into a separate file so that it can be reused outside of extended diagnostics.

### refactor(compiler-cli): move illegal template assignment check into template semantics checker
Moves the check which ensures that there are no writes to template variables into the `TemplateSemanticsChecker` to prepare for the upcoming changes.

### fix(compiler-cli): flag two-way bindings to non-signal values in templates
We have a diagnostic that reports writes to template variables which worked both for regular event bindings and two-way bindings, however the latter was broken by https://github.com/angular/angular/pull/54154 because two-way bindings no longer had a `PropertyWrite` AST.

These changes fix the diagnostic and expand it to allow two-way bindings to template variables that are signals.

### fix(compiler): handle two-way bindings to signal-based template variables in instruction generation
Updates the instruction generation for two-way bindings to only emit the `twoWayBindingSet` call when writing to template variables. Since template variables are constants, it's only allowed to write to them when they're signals. Non-signal values are flagged during template type checking.

Fixes https://github.com/angular/angular/issues/54670.

crisbeto committed:
refactor(compiler-cli): introduce template semantics checker

Introduces a new `TemplateSemanticsChecker` that will be used to flag semantic errors in the user's template. Currently we do some of this in the type check block, but the problem is that it doesn't have access to the template type checker which prevents us from properly checking cases like #54670. This pass is also distinct from the extended template checks, because we don't want users to be able to turn the checks off and we want them to run even if `strictTemplates` are disabled.

crisbeto committed:
refactor(compiler-cli): move signal identification function

Moves the function that identifies signals into a separate file so that it can be reused outside of extended diagnostics.

crisbeto committed:
refactor(compiler-cli): move illegal template assignment check into template semantics checker

Moves the check which ensures that there are no writes to template variables into the `TemplateSemanticsChecker` to prepare for the upcoming changes.

crisbeto committed:
fix(compiler-cli): flag two-way bindings to non-signal values in templates

We have a diagnostic that reports writes to template variables which worked both for regular event bindings and two-way bindings, however the latter was broken by #54154 because two-way bindings no longer had a `PropertyWrite` AST.

These changes fix the diagnostic and expand it to allow two-way bindings to template variables that are signals.

devversion commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -534,15 +534,8 @@ export class TraitCompiler implements ProgramTypeCheckAdapter {
-        }
-        diagnostics.push(...trait.handler.extendedTemplateCheck(clazz, extendedTemplateChecker));
+      if (isNamedClassDeclaration(clazz)) {
+        diagnostics.push(...checker.getDiagnosticsForComponent(clazz));
```
Won't this run for all classes that have a trait attached to them? i.e. might be a pipe for example

crisbeto replied:
Good point. I've reworked it.

devversion commented on packages/compiler-cli/src/ngtsc/typecheck/api/checker.ts:
```diff
@@ -213,6 +213,12 @@ export interface TemplateTypeChecker {
   invalidateClass(clazz: ts.ClassDeclaration): void;
 
+  /**
+   * Gets the target of a template expression, if possible.
```
Would love this doc to link to `BoundTarget.getExpressionTarget` — this only handles local refs or template variables

devversion commented on packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts:
```diff
@@ -0,0 +1,121 @@
+    // Only check top-level property reads inside two-way bindings for illegal assignments.
+    if (!(context instanceof TmplAstBoundEvent) || context.type !== ParsedEventType.TwoWay ||
+        !(ast.receiver instanceof ImplicitReceiver) ||
+        ast !== unwrapAstWithSource(context.handler)) {
```
why is this needed? Do you know in what cases it's wrapped?

crisbeto replied:
It's a bit of a gotcha since `ASTWithSource extends AST`. I don't know exactly when we use one over the other, but there are several places where we need to do this sort of unwrapping.

devversion replied:
I see!

devversion commented on packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts:
```diff
@@ -0,0 +1,121 @@
+
+    // Two-way bindings to template variables are only allowed if the variables are signals.
+    const symbol = this.templateTypeChecker.getSymbolOfNode(target, this.component);
+    if (symbol !== null && !isSignalReference(symbol)) {
```
Should we always error if `symbol === null` and we don't know, except that it's a template variable?

crisbeto replied:
I was a bit concerned that there may be cases where we don't have full type information and we would raise the error unnecessarily.

devversion replied:
Interesting. maybe worth trying in g3?

crisbeto replied:
Sure, will run a TGP now.

crisbeto replied:
Okay so even without the TGP it appears to be problematic. It was flagging one of our own `TestBed`-based tests, because it can't resolve the symbol there.

devversion replied:
Sounds good

devversion commented on packages/compiler/src/compiler_util/expression_converter.ts:
```diff
@@ -84,14 +84,24 @@ export function convertAssignmentActionBinding(
+  // If the expression is targeting a variable read, we only emit the `twoWayBindingSet` since
+  // the fallback would be attempting to write into a constant. Invalid usages will be flagged
+  // during template type checking.
+  if (ast instanceof o.ReadVarExpr) {
```
Is `ReadVarExpr` always a constant? I see it as a variable, not a constant? This seems like a large assumption from the `resolve_names` phase?

`o.variable()` creates such AST object

crisbeto replied:
It can be either a `let` or `const`. Either way we need to not re-assign it (which is what this change is doing), but just pass it to the `twoWayBindingSet` instruction.

devversion replied:
But isn't this a huge assumption? A variable is technically assignable right?

We are just assuming it to be a constant template variable because the compiler converts the LexicalReadExpr into a ReadVarExpr

I might miss something

crisbeto replied:
Template variables have always been read-only, we even have a diagnostic for it. That's why I think it's safe to assume that it would be a constant here as well.

devversion replied:
Oh yeah, I think that is true. I guess my concern is that we are assuming that `ReadVarExpr` indicates a template variable?

crisbeto replied:
Even if it's not a template variable, I think we'd want to cover it with this check. E.g. I believe that `$event` would also be a `ReadVarExpr` which shouldn't be writable either.

devversion replied:
Right, but `ReadVarExpr` could technically also be some arbitrary `var` variable (I know we don't generate this right now— but it _could_ be)

crisbeto replied:
Based on our discussion over DM: I think that this is a safe assumption to make at the moment. I've added added a comment stating that it's an assumption. Will continue as is for now.

devversion reviewed (approved):
LGTM overall, but a couple of comments

pkozlowski-opensource commented on packages/core/test/acceptance/authoring/model_inputs_spec.ts:
```diff
@@ -589,4 +589,44 @@ describe('model inputs', () => {
     expect(() => fixture.destroy()).not.toThrow();
   });
+
+  it('should support two-way binding to a signal @for loop variable', () => {
```
I would _love_ to see a test illustrating how it behaves with non-signal items in a collection.

This is also an interesting consideration of how do we see items produced by a loop - are those assignable or not (signals are, obviously)

crisbeto replied:
So a non-signal value would basically be a no-op at runtime and we have a compiler diagnostic to flag it at compile time.

crisbeto commented:
[Passing TGP](http://OCL:613817804:BASE:613885343:1709899217435:319b3de7)

crisbeto committed:
fix(compiler): handle two-way bindings to signal-based template variables in instruction generation

Updates the instruction generation for two-way bindings to only emit the `twoWayBindingSet` call when writing to template variables. Since template variables are constants, it's only allowed to write to them when they're signals. Non-signal values are flagged during template type checking.

Fixes #54670.

atscott commented:
This PR was merged into the repository by commit 5ae2bf480697b475c41ea136f08f4dca633c75b9.
